workload/schemachange: allow self deps in tableHasDependencies for DROP TABLE

spilchen opened the pull request:
This is a follow-up from #147834, which didn't have a complete fix. It was flagging DROP TABLE as having a dependency, when in fact it had a self-dependency, but it doesn't block the DROP TABLE (seen in this occurrence https://github.com/cockroachdb/cockroach/issues/147514#issuecomment-2948236246)

 
This change updates tableHasDependencies to optionally include self-referencing dependencies when skipSelfRef is false. This is needed for DROP TABLE, where self-dependencies (e.g., from triggers) are safe to ignore because they are cleaned up as part of the drop.

In contrast, RENAME TABLE must still block on these, since the dependent reference would not be updated by the rename.

The function now takes a skipSelfRef flag to distinguish these cases, and excludes only self-dependencies when skipSelfRef is true.

Informs #147514

Epic: None
Release note: none

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147906#-:-OS4CJLV8Q5-VvEePhWA:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong and @herkolategan)



<!-- Sent from Reviewable.io -->

spilchen committed:
workload/schemachange: allow self deps in tableHasDependencies for DROP TABLE

This change updates tableHasDependencies to optionally include
self-referencing dependencies when skipSelfRef is false. This is
needed for DROP TABLE, where self-dependencies (e.g., from triggers)
are safe to ignore because they are cleaned up as part of the drop.

In contrast, RENAME TABLE must still block on these, since the
dependent reference would not be updated by the rename.

The function now takes a skipSelfRef flag to distinguish these cases,
and excludes only self-dependencies when skipSelfRef is true.

Informs #147514

Epic: None
Release note: none

spilchen commented:
TFTR!

bors r+
